src/lib.rs

```diff
@@ -0,0 +1,1210 @@
+        }
+        write!(fmt, "]")?;
+        Ok(())
+    }
+}
+
+impl<T: Debug + One + PrimInt + Zero> Extend<bool> for Vob<T> {
+    fn extend<I: IntoIterator<Item=bool>>(&mut self, iterable: I) {
+        let iterator = iterable.into_iter();
+        let (min, _) = iterator.size_hint();
+        self.reserve(min);
+        for e in iterator {
+            self.push(e)
+        }
+    }
+}
+
+impl FromIterator<bool> for Vob<usize> {
+    /// Create a Vob from an iterator.
+    ///
+    /// # Examples
+    /// ```
+    /// use std::iter::FromIterator;
+    /// use vob::Vob;
+    /// let v = Vob::from_iter(vec![true, false]);
+    /// assert_eq!(v, Vob::from_iter(vec![true, false]));
+    /// ```
+    fn from_iter<I: IntoIterator<Item=bool>>(iter: I) -> Self {
+        let mut v = Vob::new();
+        v.extend(iter);
+        v
+    }
+}
+
+
+// This is based on the `BitVec` approach to indices. It's clearly a horrible way of doing things,
+// but until `IndexGet` is implemented, we're stuck.
+static TRUE: bool = true;
+static FALSE: bool = false;
+
+impl<T: Debug + One + PrimInt + Zero> Index<usize> for Vob<T> {
+    type Output = bool;
+
+    fn index(&self, index: usize) -> &bool {
+        match self.get(index) {
+            Some(true) => &TRUE,
+            Some(false) => &FALSE,
+            None => panic!("index out of bounds: the len is {} but the index is {}", self.len, index)
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct Iter<'a, T: 'a> {
+    vob: &'a Vob<T>,
+    range: Range<usize>,
+}
+
+impl<'a, T: Debug + One + PrimInt + Zero> Iterator for Iter<'a, T> {
+    type Item = bool;
+
+    fn next(&mut self) -> Option<bool> {
+        self.range.next()
+                  .map(|i| self.vob.get(i).unwrap())
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.range.size_hint()
+    }
+}
+
+impl<'a, T: Debug + One + PrimInt + Zero> DoubleEndedIterator for Iter<'a, T> {
+    fn next_back(&mut self) -> Option<bool> {
+        self.range.next_back()
+                  .map(|i| self.vob.get(i).unwrap())
+    }
+}
+
+impl<'a, T: Debug + One + PrimInt + Zero> ExactSizeIterator for Iter<'a, T> { }
+
+impl<'a, T: Debug + One + PrimInt + Zero> IntoIterator for &'a Vob<T> {
+    type Item = bool;
+    type IntoIter = Iter<'a, T>;
+
+    fn into_iter(self) -> Iter<'a, T> {
+        self.iter()
+    }
+}
+
+#[derive(Clone)]
+pub struct IterSetBits<'a, T: 'a> {
+    vob: &'a Vob<T>,
+    range: Range<usize>,
+}
+
+impl<'a, T: Debug + One + PrimInt + Zero> Iterator for IterSetBits<'a, T> {
+    type Item = usize;
+
+    fn next(&mut self) -> Option<usize> {
+        if let Some(mut i) = self.range.next() {
+            for b in block_offset::<T>(i) .. blocks_required::<T>(self.vob.len) {
+                let v = self.vob.vec[b];
+                if v != T::zero() {
+                    let i_off = i % bits_per_block::<T>();
+                    let tz = (v >> i_off).trailing_zeros() as usize;
```

Comment:
Just to check, this function is fully generic? It will work for other block types?

Reply:
Erp, good point. In practise, yes... but I agree, it doesn't look good. It might be difficult to do better though. Let me try.

Reply:
OK, this is in fact safe; but we could probably use `u32` for `bits_per_block` as the `trailing_zeros` function does. That upsets a lot of code though because `num_bits` is `usize`, which means that any operation involving the two types requires `as`. I'm not sure it's worth fixing.

Reply:
ok